cloudwatch_logs_common/test/log_batch_test.cpp

```diff
@@ -0,0 +1,246 @@
+  Aws::DataFlow::UploadStatus attemptPublish(std::list<Aws::CloudWatchLogs::Model::InputLogEvent> &data) override {
+    last_upload_status = Publisher::attemptPublish(data);
+    {
+      this->notify();
+    }
+    return last_upload_status;
+  }
+
+  Aws::DataFlow::UploadStatus publishData(std::list<Aws::CloudWatchLogs::Model::InputLogEvent>&) override {
+
+    if (force_failure) {
+      return Aws::DataFlow::UploadStatus::FAIL;
+
+    } else if (force_invalid_data_failure) {
+      return Aws::DataFlow::UploadStatus::INVALID_DATA;
+
+    } else {
+      return Aws::DataFlow::UploadStatus::SUCCESS;
+    }
+  }
+
+  bool force_failure;
+  bool force_invalid_data_failure;
+  Aws::DataFlow::UploadStatus last_upload_status;
+};
+
+class LogBatchTest : public ::testing::Test {
+public:
+  void SetUp() override
+  {
+      test_publisher = std::make_shared<TestPublisher>();
+      batcher = std::make_shared<LogBatcher>();
+
+      //  log service owns the streamer, batcher, and publisher
+      cw_service = std::make_shared<LogService>(test_publisher, batcher);
+
+      stream_data_queue = std::make_shared<TaskObservedQueue<LogCollection>>();
+      queue_monitor = std::make_shared<Aws::DataFlow::QueueMonitor<TaskPtr<LogCollection>>>();
+
+      // create pipeline
+      batcher->setSink(stream_data_queue);
+      queue_monitor->addSource(stream_data_queue, Aws::DataFlow::PriorityOptions{Aws::DataFlow::HIGHEST_PRIORITY});
+      cw_service->setSource(queue_monitor);
+
+      cw_service->start(); //this starts the worker thread
+      EXPECT_EQ(Aws::DataFlow::UploadStatus::UNKNOWN, test_publisher->getLastUploadStatus());
+  }
+
+  void TearDown() override
+  {
+      if (cw_service) {
+        cw_service->shutdown();
+        cw_service->join();
+      }
+    //std::experimental::filesystem::path storage_path(options.storage_directory);
+    //std::experimental::filesystem::remove_all(storage_path);
+  }
+
+protected:
+  FileManagerStrategyOptions options{"test", "log_tests/", ".log", 1024*1024, 1024*1024};
+
+  std::shared_ptr<TestPublisher> test_publisher;
+  std::shared_ptr<LogBatcher> batcher;
+  std::shared_ptr<LogService> cw_service;
+
+  std::shared_ptr<TaskObservedQueue<LogCollection>> stream_data_queue;
+  std::shared_ptr<Aws::DataFlow::QueueMonitor<TaskPtr<LogCollection>>>queue_monitor;
+};
+
+auto log_comparison = [](const LogType & log1, const LogType & log2)
+  { return log1.GetTimestamp() < log2.GetTimestamp(); };
+
+/**
+ * Test File Manager
+ */
+class TestLogFileManager : public FileManager<LogCollection>, public Waiter
+{
+public:
+
+    TestLogFileManager() : FileManager(nullptr) {
+      written_count.store(0);
+    }
+
+    void write(const LogCollection & data) override {
+      last_data_size = data.size();
+      written_count++;
+      this->notify();
+    };
+
+    //test that the readBatch function works with 24 hour interval
+    FileObject<LogCollection> readBatch(size_t batch_size) override {
```

Comment:
Is this method copied from the implementing class? If so, what happens if the method in the implementation changes, but the test does not?

I would mock the file reading specifics, i.e., `readLine(...)` so you can test your method directly while supplying the fake data through that method (as you have done below).